### PR TITLE
feature(wyh): on policy ppg algorithm and procgen configs

### PR DESCRIPTION
**Pull request by Weiyuhong-1998**
## Description
Add on policy ppg algorithm
Add IMPALA CNN Encoder
Add Procgen configs

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/policy/command_mode_policy_instance.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from .sqn import SQNPolicy
 from .ppo import PPOPolicy, PPOOffPolicy
 from .offppo_collect_traj import OffPPOCollectTrajPolicy
-from .ppg import PPGPolicy
+from .ppg import PPGPolicy, OnPPGPolicy
```
Comment (PaParaZz1) [inline ding/policy/command_mode_policy_instance.py:16]: rename to `PPGOnPolicy`

File: ding/model/template/vac.py [hunk 2]
```diff
@@ -73,7 +76,10 @@ def __init__(
                     format(obs_shape)
                 )
             self.share_encoder = share_encoder
-            if self.share_encoder:
+            if self.share_encoder and self.use_impala_cnn_encoder:
```
Comment (PaParaZz1) [inline ding/model/template/vac.py:79]: don't implement it like this, both share/no-share cases should be supported

File: ding/torch_utils/network/impalacnn_encoder.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    """
+    Product of a sequence of integers
+    """
+    out = 1
+    for x in xs:
+        out *= x
+    return out
+
+
+def zeros(*args, **kwargs):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/impalacnn_encoder.py:23]: directly use `torch.zeros`

File: ding/torch_utils/network/impalacnn_encoder.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    nn.Conv2d but with normalized fan-in init
+    """
+    out = nn.Conv2d(*args, **kwargs)
+    out.weight.data *= scale / out.weight.norm(dim=(1, 2, 3), p=2, keepdim=True)
+    if kwargs.get("bias", True):
+        out.bias.data *= 0
+    return out
+
+
+def flatten_image(x):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/impalacnn_encoder.py:51]: use `ding.torch_utils.Flatten`

File: ding/torch_utils/network/impalacnn_encoder.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    """
+    Flattens last three dims
+    """
+    *batch_shape, h, w, c = x.shape
+    return x.reshape((*batch_shape, h * w * c))
+
+
+def sequential(layers, x, *args, diag_name=None):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/impalacnn_encoder.py:59]: use `torch.nn.Sequential`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/vac.py [hunk 6]
```diff
@@ -35,6 +36,7 @@ def __init__(
         fixed_sigma_value: Optional[int] = 0.3,
         bound_type: Optional[str] = None,
         encoder: Optional[torch.nn.Module] = None,
+        use_impala_cnn_encoder: bool = False,
```
Comment (PaParaZz1) [inline ding/model/template/vac.py:39]: rename to `impala_cnn_encoder`, don't use `use_xxx` or `is_xxx`

File: ding/model/common/encoder.py [hunk 7]
```diff
@@ -153,3 +160,196 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        for outchan in chans:
+            stack = CnnDownStack(curshape[0], nblock=nblock, outchan=outchan, scale=s, **kwargs)
+            self.stacks.append(stack)
+            curshape = stack.output_shape(curshape)
+        self.dense = NormedLinear(intprod(curshape), outsize, scale=1.4)
+        self.outsize = outsize
+        self.final_relu = final_relu
+
+    def forward(self, x):
+        x = x.to(dtype=torch.float32) / self.scale_ob
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:332]: remove `to` operation, it is unnecessary here

File: ding/model/template/vac.py [hunk 8]
```diff
@@ -74,16 +76,25 @@ def __init__(
                 )
             self.share_encoder = share_encoder
             if self.share_encoder:
-                self.encoder = encoder_cls(
-                    obs_shape, encoder_hidden_size_list, activation=activation, norm_type=norm_type
-                )
+                if self.impala_cnn_encoder:
+                    encoder_cls = ImpalaEncoder
+                    self.encoder = encoder_cls(obs_shape)
```
Comment (PaParaZz1) [inline ding/model/template/vac.py:81]: just use `self.encoder = ImpalaEncoder(obs_shape)`

File: ding/policy/ppg.py [hunk 9]
```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
...
+           | ``lambda``                                 | of bias and variance(1-step td and mc)
+        == ==================== ======== ============== ======================================== =======================
+    """
+    config = dict(
+        # (str) RL policy register name (refer to function "POLICY_REGISTRY").
+        type='ppg',
+        # (bool) Whether to use cuda for network.
+        cuda=False,
+        # (bool) Whether the RL algorithm is on-policy or off-policy. (Note: in practice PPO can be off-policy used)
+        on_policy=False,
```
Comment (PaParaZz1) [inline ding/policy/ppg.py:89]: `on_policy=True`

File: ding/policy/ppg.py [hunk 10]
```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
...
+                    value_output['value'], value_data['value'], return_, value_data['weight']
+                )
+                value_loss = self._value_weight * ppo_value_error(value_error_data, self._clip_ratio)
+                self._optimizer_aux_critic.zero_grad()
+                value_loss.backward()
+                self._optimizer_aux_critic.step()
+
+        data['return_'] = data['return']
+
+        self._aux_memories.append(copy.deepcopy(data))
```
Comment (PaParaZz1) [inline ding/policy/ppg.py:279]: test whether we can remove `deepcopy` here

File: ding/policy/ppg.py [hunk 11]
```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
...
+                )
+                ppo_policy_loss, ppo_info = ppo_policy_error(policy_error_data, self._clip_ratio)
+                policy_loss = ppo_policy_loss.policy_loss - self._entropy_weight * ppo_policy_loss.entropy_loss
+                self._optimizer_ac.zero_grad()
+                policy_loss.backward()
+                self._optimizer_ac.step()
+
+        for epoch in range(self._cfg.learn.critic_epoch_per_collect):
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                value_data = batch
```
Comment (PaParaZz1) [inline ding/policy/ppg.py:261]: remove this line, just use `for value_data in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):`

File: ding/policy/ppg.py [hunk 12]
```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
...
+            gamma=self._gamma,
+            gae_lambda=self._gae_lambda,
+            cuda=False,
+        )
+        if self._value_norm:
+            for i in range(len(data)):
+                data[i]['value'] /= self._running_mean_std.std
+
+        for i in range(len(data)):
+            data[i].pop('next_obs')
```
Comment (PaParaZz1) [inline ding/policy/ppg.py:446]: why add this `pop` operation

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/ppg.py (hunk 10 above)
Comment (Weiyuhong-1998) [reply ding/policy/ppg.py:279]: No, learn_aux was not always 1

File: dizoo/procgen/bigfish/entry/bigfish_ppg_config.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+bigfish_ppg_default_config = dict(
+    exp_name='bigfish_ppg_seed0',
+    env=dict(
+        is_train=True,
+        collector_env_num=64,
+        evaluator_env_num=10,
+        n_evaluator_episode=50,
+        stop_value=40,
+        manager=dict(
```
Comment (PaParaZz1) [inline dizoo/procgen/bigfish/entry/bigfish_ppg_config.py:11]: enable `shared_memory=True`

File: dizoo/procgen/bigfish/entry/bigfish_ppg_config.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            critic_epoch_per_collect=1,
+            value_norm=True,
+            batch_size=16384,
+            value_weight=0.5,
+            entropy_weight=0.01,
+            clip_ratio=0.2,
+            aux_freq=1,
+        ),
+        collect=dict(n_sample=16384, ),
+        eval=dict(evaluator=dict(eval_freq=96, )),
```
Comment (PaParaZz1) [inline dizoo/procgen/bigfish/entry/bigfish_ppg_config.py:37]: why `eval_freq=96`, use a bigger value?

File: dizoo/procgen/coinrun/envs/coinrun_env.py [hunk 15]
```diff
@@ -1,31 +1,33 @@
 from typing import Any, List, Union, Optional
```
Comment (PaParaZz1) [inline dizoo/procgen/coinrun/envs/coinrun_env.py:1]: can you merge `coinrun`, `maze`, `bigfish` into a unified `ProcgenEnv` with different env_id args

File: ding/model/common/encoder.py [hunk 16]
```diff
@@ -159,3 +162,188 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    nn.Conv2d but with normalized fan-in init
+    """
+    out = nn.Conv2d(*args, **kwargs)
+    out.weight.data *= scale / out.weight.norm(dim=(1, 2, 3), p=2, keepdim=True)
+    if kwargs.get("bias", True):
+        out.bias.data.zero_()
+    return out
+
+
+class Encoder(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:201]: remove this base class

File: ding/model/common/encoder.py [hunk 17]
```diff
@@ -159,3 +162,188 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        *batch_shape, h, w, c = x.shape
+        x = x.reshape((*batch_shape, h * w * c))
+        x = F.relu(x)
+        x = self.dense(x)
+        if self.final_relu:
+            x = torch.relu(x)
+        return x
+
+
+class ImpalaEncoder(Encoder):
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:338]: merge this class into `ImpalaCNN` and rename to `ImpalaConvEncoder`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/ppg.py [hunk 18]
```diff
@@ -34,6 +34,585 @@ def create_shuffled_dataloader(data, batch_size):
     return DataLoader(ds, batch_size=batch_size, shuffle=True)
 
 
+@POLICY_REGISTRY.register('ppg_onpolicy')
+class PPGOnPolicy(Policy):
```
Comment (PaParaZz1) [inline ding/policy/ppg.py:38]: rename these two classes, `PPGPolicy` and `PPGOffPolicy` , and modify related parts

**Review by PaParaZz1 — APPROVED** [review]
